Refactors kinetic crusher damage tracking, cleans up and splits up `kinetic_crusher.dm`

Sealed101 opened the pull request:
## About The Pull Request

This PR aims to move crusher damage tracking from the crusher's `attack()` and `afterattack()` and crusher projectile's `on_hit()` to be handled by a component, offloading janky and unreliable tracking to a hopefully more responsive and more easily expandable system.

The current crusher damage tracking is unreliable as described in #73887, which to be completely honest was caused by a single `..()` instead of a `. = ..()`. However, just that was not enough to completely fix the tracking, as the last hit from a crusher mark detonation would not count towards this in some cases - it would proc after crusher's `attack()`, and the target could be by that point dead. ~~Though, this is currently the case still, I aim to move that to the crusher mark's effect or something.~~ With the current implementation, the crusher mark pops _after_ the attack & trophy attack effects yet _before_ any trophy mark detonation effects, which should make it more stable (in testing it was at least lol)

With the component, I aim to expand the crusher damage tracking to also increment damage from crusher trophies' effects, like making the destabilizer projectile deal damage, or spell effects like hierophant blasts or drake spike. This would also mean that such effects could also apply the crusher damage tracker effect, which they currently cannot.

I couldn't wrangle `apply_damage()` or similar into cooperating, so this system operates with raw damage values without passing an armor check, if any. This could mean better chances for miners to get their trophies, but I don't think it would cause that much more trophies. We'll see, I hope to test that further when it is more stable (see To-Do)
- During testing, the only blatant outliers were megafauna getting struck with damage from ash drake or hierophant trophies. Megas have innate fire armor, and since I don't have any control on the actual damage the mob has sustained, I am using the raw damage for sending these values. TBQH I don't have much issue with that, just more trophy rates for crusher+PKA miner teams.

#### This PR does not make multiple crushers work off of each other's crusher marks. However, there could be room for maybe making the marks have customizable color so players can differentiate better between their marks.

There were also a few QoL things, like making the examine block display the trophy attached to the crusher, as well as formatting the effect of said trophy a little better; making a balloon message when you attach a trophy, be it with slapping the trophy onto the crusher or slapping the crusher onto the trophy (#76212 hi sexrule)... probably one or two more that I've forgotten about since I've been cooking this for a few months now.
Oh, attacking a goliath or any other mob with melee range revenge ability with a crusher in melee works now.
Double oh — the file split. The file was fuckhueg, mostly because it has had trophies in it. A new folder has been created, and the file has been split into `_kinetic_crusher.dm`, `_trophy.dm`, `destabilizer_projectile.dm` and a subfolder for trophy types separated by mob origin and elite or megafauna status.

## To-Do

- [x] (components fixed this) fix elements overriding each others' damage (like firing a projectile would override crusher melee damage)
- [x] remove debug shit
- [x] more code cleanup?
- [x] make removing trophies a tgui list/radial list to drop one at a time?
- [x] make trophy effects hook onto signals on the crusher? would be better instead of going through a list of all trophies
- [ ] customizable destabilizer projectile/crusher mark field colors?
- [ ] ~~fix trophies that affect force stacking incorrectly~~ not really worth the trouble? i think?
- [ ] logging and unit testing question mark?

## Why It's Good For The Game

- Closes #73887
- Closes #76212 post-mortem, also prevents further mix-ups of this nature

More consistent crusher damage tracking, which results in more trophy drop rates, which results in happier miners.
Clearer descriptions of the trophies attached is also good, easier to parse what you've glued onto your mining stick.

## Changelog

:cl:
refactor: refactored kinetic crusher damage tracking to be more robust. there should be much less weird losses of crusher damage, which should mean more consistent trophy drop rates.
refactor: this also makes crusher trophies' effects increment crusher damage done (as well as allowing them to start a crusher damage tracker on a victim), i.e. a hiero blast trail's damage will also count towards the total crusher damage dealt to a mob.
refactor: crusher trophy effects are now signal-based, hopefully making them more responsive. this may have changed the order of how the effects are triggered, so please report any inconsistencies with the previous behavior
qol: attaching a trophy to a crusher makes a balloon alert (both attaching a trophy to a crusher and clicking a trophy with a crusher)
qol: trophies attached to a crusher are listed better, with a trophy's icon in the crusher's examine block and some formatting
qol: removing a trophy from a crusher uses a radial menu, with tooltips showing the trophy's effect
qol: crusher detonation/backstab damage reworded to be less misleading - it's a bonus damage instance separate from your melee attack, not a hugelarge boosted attack of the crusher
fix: fixed demon claws trophy healing effect working on dead mobs
fix: fixed goliath melee tentacles revenge ability not triggering on crusher attacks (this also fixes their ranged tentacles revenge but that's more of a problem with how they listen to "attacks")
fix: Kilo Whiteship turrets now share the same type. Don't ask how is this relevant to the crusher changes.
/:cl:

Sealed101 committed:
blooning
that's it, that's the PR...
...is what i would have said if i didn't fuck up yet another branch merge conflcit

Sealed101 committed:
actually let me give the code a pass
dmdoc, singlevars, some other bullshit
i'll probably move the trophies into a different file
why the hell are some of the trophies away from this file
SERIOUSLY WHO SPELLED IT WRONG???

(i know but i won't tell you)

Sealed101 committed:
5 reasons why you should test your code
1. the first commit of this PR (in another timeline (this branch revival amended that slip-up))
2-5 - IOU reasons
____
this commit puts the crusher and the trophies in their own folder
separates the crusher and the trophies into their own files
moves all trophies to a singular file
hopefully completes dmdoc and var name stuff

Sealed101 committed:
code's in the file where i need it
except i misplaced bileworm lol but it's in trophies now (and i didn't rectify this on this new branch, sue me)
legionnaire ribcage makes sounds when activated/used in-hand
same for goliath broodmother's tongue
good times have been had by all
trophy listing on crusher examine is a bit clearer now
some span changes, some desc changes to reflect the trophy nature
some of the trophies use bonus_value for their effects, adminbus rejoice
lowers volume (ow)

Sealed101 committed:
REMOVES SOUL (again)
kenetic -> kinetic

Sealed101 committed:
small fix (i don't think it's dumb anymore)
prevent unwielded crushers from popping a mark because lol

Sealed101 committed:
BITES ZA DUSTO
(hopefully) resolves conflicts that caused the end of the universe for the previous branch

Sealed101 committed:
Merge branch 'master' into crusher-trophies-cleanup

Sealed101 committed:
basic trophies go away to your trophy corner

Sealed101 committed:
some small things
crusher desc reworded to display bonus damage for the mark detonation/backstab - it was calculating it with force of the crusher which when unwielded is 0
fixes bubbles claws and wendigo horn making unwielded force equal to wielded - though the issue of them not stacking properly still exists

Sealed101 committed:
crusher damage tracker applying  element
currently applied to the crusher, the destabilizer projectile and the bileworm trophy acid spew projectiles

Sealed101 committed:
hierophant trophy applies crusher damage tracker
or rather, the blasts of the trophy's chaser

Sealed101 committed:
broodmother trophy applies crusher tracker
also fixes it affecting the user and their faction members

Sealed101 committed:
drake trophy applies crusher damage tracker
yeah that's about it

Sealed101 committed:
legionnaire trophy applies crusher damage tracker

Sealed101 committed:
splits up the files further
mainly the trophies file because thinking back it was a mess to navigate
crusher, the projectile and the base trophy item stay on top

Sealed101 committed:
Merge branch 'master' into crusher-trophies-cleanup

Sealed101 committed:
crusher_damage_applicant -> crusher_damage_ticker
because we're going to do the funny now.

Sealed101 committed:
attempt to move damage increment to the element
it's broken because something causes several sources of crusher damage to overlap and override each other's damage
like shooting a projectile will override crusher's damage to pass
this is cringe and i don't know what's causing it, but just smacking shit with the crusher with no effects works fine
does not move mark detonation

also fixes goliaths not triggering their revenge attacks on crusher melee
also makes goliath trophy damage a ticker

also adds signals to twohanded parent item force updates
may not be the best solution to the problem of the crusher changing its damage on wield/unwield but we'll see
still got to fix how bubblegum/wendigo damage buffs are handled

Sealed101 commented:
this might have been saner if `apply_damage()` could have had a better way to implement differentiation between damage sources but i didn't find anything of use

pls help

Sealed101 committed:
nullcheck tentacle caster because reasons
thank you c&d
also removes the dumbass debug thing because its failing ci and is not helping me with finding what the fuck is causing the element to override other element instances' damage_to_increment

Sealed101 committed:
makes it a component intead
it magically fixed everything, incredible

Sealed101 committed:
proper goliath trophy damage
drake trophy checks faction
non-modified destabilizer doesn't start tracking

Sealed101 committed:
moves mark detonation to the component

Sealed101 committed:
adds combat logging to the spells

Sealed101 commented:
it works :tm:
the only outlier i think are spell-type damage sources that use `apply_damage()`, which is the hiero trophy. since some fauna has burn damage modifiers, they may accrue more crusher damage than their max health. 
also not sure about the validity of adding this component to temporary effects like tendrils or hiero blast in the first place, if this can become an issue let me know
there was also an issue with crusher tracker failing to apply on stuff, will need to replicate

otherwise uh review-ready sort of? still need to iron out some stuff and decide what things on the to-do list are worth doing.

Sealed101 committed:
radial menu for removing trophies from a crusher
bit saner crusher attackby() and Exited()
also uncomments an afterattack fix

Sealed101 committed:
more trophy desc formatting

Sealed101 commented:
radial menu trophy removing implemented, with the info buttons showing the effect of the trophy in case you want to double check the effect
lmk if a `Dump all` option would be appropriate

![изображение](https://github.com/tgstation/tgstation/assets/75863639/5d0d9b5d-dfa6-42e8-a928-e8171ccb5e5c)

Sealed101 committed:
removes traces of attempts to fix stacking
i remembered that wendigo and bubblegum have such a slim chance of spawning on icebox that it barely matters
also i'm lazy yeah

Sealed101 committed:
generic post (un)wield signals
thanks fikou

Sealed101 committed:
removes debug stuff + adds dmdoc to stuff

Sealed101 committed:
Merge branch 'master' into crusher-trophies-cleanup

Sealed101 committed:
why did dm lang server not catch this
compiling was taking twice as long as usual and didn't finish, so i didn't see it there either
the struggle continues

Sealed101 committed:
this motherfucker was failing map linters
mapping varedits are a grave sin

Sealed101 committed:
Merge branch 'master' into crusher-trophies-cleanup

Jacquerel commented on code/modules/mining/equipment/kinetic_crusher/trophies/lavaland_mob_trophies.dm:
```diff
@@ -0,0 +1,208 @@
+	return "mark detonation to make the next destabilizer shot deal <b>[bonus_value]</b> damage"
+
+/obj/item/crusher_trophy/blaster_tubes/magma_wing/on_projectile_fire(obj/projectile/destabilizer/marker, mob/living/user)
+	if(deadly_shot)
```
```suggestion
	if(!deadly_shot)
		return
```
I almost wonder if for this & the other one you could just replace the marker shot with a different projectile somehow... this is probably fine though.

Jacquerel commented on code/modules/mining/equipment/kinetic_crusher/trophies/lavaland_mob_trophies.dm:
```diff
@@ -0,0 +1,208 @@
+/**
+ * Lobstrosity
+ * Detonating a mark causes the victim to stagger for 1 second, which slows down movement
+ * and more than doubles the cooldown for the victim's ranged attack (if it's a simple hostile mob).
```
...so it largely also doesn't do anything 😔 
making these work doesnt necessarily need to be in this PR but it's sad

Jacquerel commented on code/modules/mining/equipment/kinetic_crusher/trophies/lavaland_mob_trophies.dm:
```diff
@@ -0,0 +1,208 @@
+	var/datum/action/cooldown/mob_cooldown/projectile_attack/dir_shots/spewlet/ability
+
+/obj/item/crusher_trophy/bileworm_spewlet/effect_desc()
+	return "mark detonation to <b>launch projectiles</b> in cardinal directions on a <b>10</b> second cooldown"
```
The reason that I think that `bonus_value` is pointless is that people stop bothering to use it

Sealed101 replied:
this trophy is a sin and armhulen knew that when making it

Jacquerel commented on code/modules/mob/living/basic/lavaland/goliath/tentacle.dm:
```diff
@@ -41,16 +47,23 @@
 /obj/effect/goliath_tentacle/proc/grab()
 	for (var/mob/living/victim in loc)
-		if (victim.stat == DEAD || HAS_TRAIT(victim, TRAIT_TENTACLE_IMMUNE))
+		if (victim.stat == DEAD || HAS_TRAIT(victim, TRAIT_TENTACLE_IMMUNE) || caster?.faction_check_mob(victim))
```
Was there a faction check previous to my rework? I kind of _like_ goliaths being able to friendly fire other mobs.

Sealed101 replied:
it was here before but we can just have goliaths not pass caster maybe

Jacquerel commented on code/modules/mob/living/basic/space_fauna/netherworld/migo.dm:
```diff
@@ -34,7 +34,7 @@
 /mob/living/basic/migo/Initialize(mapload)
 	. = ..()
-	migo_sounds = list('sound/items/bubblewrap.ogg', 'sound/items/change_jaws.ogg', 'sound/items/crowbar.ogg', 'sound/items/drink.ogg', 'sound/items/deconstruct.ogg', 'sound/items/carhorn.ogg', 'sound/items/change_drill.ogg', 'sound/items/dodgeball.ogg', 'sound/items/eatfood.ogg', 'sound/items/megaphone.ogg', 'sound/items/screwdriver.ogg', 'sound/items/weeoo1.ogg', 'sound/items/wirecutter.ogg', 'sou ...
+	migo_sounds = list('sound/items/bubblewrap.ogg', 'sound/items/change_jaws.ogg', 'sound/items/crowbar.ogg', 'sound/items/drink.ogg', 'sound/items/deconstruct.ogg', 'sound/items/carhorn.ogg', 'sound/items/change_drill.ogg', 'sound/items/dodgeball.ogg', 'sound/items/eatfood.ogg', 'sound/items/megaphone.ogg', 'sound/items/screwdriver.ogg', 'sound/items/weeoo1.ogg', 'sound/items/wirecutter.ogg', 'sou ...
```
~~can you alphabetise this list~~

Jacquerel commented on code/modules/mob/living/basic/lavaland/goliath/tentacle.dm:
```diff
@@ -41,16 +47,23 @@
+		if(caster)
+			log_combat(caster, victim, "grabbed with a ", src, trophy_spawned ? "crusher-spawned" : "")
+		if(trophy_spawned)
+			SEND_SIGNAL(src, COMSIG_CRUSHER_SPELL_HIT, victim, caster, damage_dealt)
```
Hmm, we are sending this signal from a lot of places, which is against our usual design for signals.
I understand _why_ we are though and it would be annoying to have a new signal for every one of these. Tricky.

Sealed101 commented on code/modules/mining/equipment/kinetic_crusher/_trophy.dm:
```diff
@@ -0,0 +1,63 @@
+	icon = 'icons/obj/mining_zones/artefacts.dmi'
+	icon_state = "tail_spike"
+	///Generic bonus to X var; if the trophy has a bonus effect, this is how much that effect is
+	var/bonus_value = 10
```
i stand by having bonus values as vars because that makes adminbus funnier + who knows maybe i will actually do something about trophy upgrading in the future (after lavaland is basicmob'd approximately)

Jacquerel replied:
I don't think a generic ability to upgrade trophies would make any sense because these values are treated as probabilities and seconds and "other" and aren't even utilised by all implementations. You can't add or multiply the value because this would have a wildly different effect on different items.

If you want these to be available for admin abuse then add them as editable vars on the individual item, allowing you to give them a descriptive name for what the admin is actually editing.

Sealed101 committed:
first review pass
comment changes
early returns
crusher material cost rounded down

Sealed101 committed:
second review pass
goliath actions don't pass a caster argument to the tentacles — lavaland has the mp_friendlyfire 1
legionnaire trophy skull spawn is a proc
goliath broodmother trophy use effect now uses a status effect

Sealed101 committed:
third review pass
gets rid of bonus_value, replaces it with other adminbus compliant vars
denied_types list
better time display on trophy effects
logging fixes (legions qdel on death moment)

Sealed101 committed:
signals the trophy effects
some more sound/desc fixes along the way
demon claws damage modifier still doesn't want to work for some goddamn reason
however i did make them not heal you when attacking a dead mob so there's that
i should really test this again because the order of effects changed with the signals used a little, but oh well~
maybe adding a logging file would help

Sealed101 committed:
Merge branch 'master' into crusher-trophies-cleanup
this will severely affect the trout population

Sealed101 committed:
Merge branch 'master' into crusher-trophies-cleanup
banishes brimdeon trophy to the trophy enclosure

Sealed101 committed:
adds a sound effect to goliath trophy

Sealed101 committed:
more volume to the goliath trophy
also some tampering with the crusher mark (legions keep runtiming because they're qdel_on_death)

Sealed101 committed:
saner custom_materials for the crusher
i mean uh yeah

Jacquerel commented on code/datums/components/crusher_damage_ticker.dm:
```diff
@@ -0,0 +1,96 @@
+
+///Tries to apply a crusher damage status effect. Returns either a new one or an already existing status effect.
+/datum/component/crusher_damage_ticker/proc/try_apply_damage_tracker(atom/target)
+	if(QDELETED(target))
+		return
+	if(!isliving(target))
```
```suggestion
	if(QDELETED(target) || !isliving(target))
```

Jacquerel commented on code/datums/components/crusher_damage_ticker.dm:
```diff
@@ -0,0 +1,96 @@
+	if(has_tracker)
+		return has_tracker
+	var/new_apply_tracker = living_target.apply_status_effect(/datum/status_effect/crusher_damage)
+	if(!new_apply_tracker)
```
```suggestion
	if(isnull(new_apply_tracker))
```

Jacquerel commented on code/datums/status_effects/debuffs/debuffs.dm:
```diff
@@ -326,37 +326,70 @@
+		combined_damage += hammer_synced.backstab_bonus
+		playsound(user, 'sound/weapons/kinetic_accel.ogg', 80, TRUE)
+
+	SEND_SIGNAL(src, COMSIG_CRUSHER_SPELL_HIT, owner, user, combined_damage)
```
I thought this was sent by trophies triggering weird effects, why is it also on the mark detonation?

Jacquerel replied:
plus sending it on `src` on a status effect is pretty sus

Jacquerel commented on code/modules/mining/equipment/kinetic_crusher/trophies/lavaland_megafauna_trophies.dm:
```diff
@@ -0,0 +1,183 @@
+		addtimer(CALLBACK(src, PROC_REF(pushback), victim, user), 1) //no free backstabs, we push AFTER module stuff is done
+		log_combat(user, target, "repelled with a trophy ", src, "crusher damage")
+		victim.adjustFireLoss(blast_damage, forced = TRUE)
+		SEND_SIGNAL(src, COMSIG_CRUSHER_SPELL_HIT, victim, user, blast_damage)
```
Make a proc `/obj/item/crusher_trophy/report_spell_damage` or some such and send the signal from there, calling it from each trophy with one of these effects. It's better than sending the same signal from a bunch of different procs.

What you do for the tentacles and hierophant chaser... 🤷 
Maybe you can think of an alternate solution. Perhaps you can just pass an appropriate existing triggering signal into the component as an argument?

Jacquerel replied:
Or register a signal on the spawned effect from the trophy when it is spawned, which calls `report_spell_damage` when an appropriate signal is sent by the effect.

Sealed101 committed:
Merge branch 'master' into crusher-trophies-cleanup
e

Sealed101 committed:
brainfart

Sealed101 commented:
my recent cold has knocked the wind out of me, taking a ~week to rest. will come back to this in a new PR
